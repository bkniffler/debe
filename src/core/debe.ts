import {
  IObserverCallback,
  IItem,
  IGetItem,
  IInsertItem,
  IQueryInput,
  IInsertInput,
  IUnlisten,
  IInsert,
  IDebeUse
} from './types';
import { DebeDispatcher } from './dispatcher';
import { ensureQuery } from './sanitize';

export class Debe<TBase = IItem, TDispatcher = DebeDispatcher<TBase>> {
  dispatcher: TDispatcher & DebeDispatcher<TBase>;
  initializing: Promise<void>;
  isInitialized = false;
  constructor(dispatcher: TDispatcher & DebeDispatcher<TBase>) {
    this.dispatcher = dispatcher;
    dispatcher.db = this;
  }
  async close() {
    await this.dispatcher.close();
  }
  async initialize() {
    if (!this.initializing) {
      this.initializing = this.dispatcher.initialize();
    }
    await this.initializing;
    this.isInitialized = true;
    return this;
  }
  use<T = IItem>(collection: string): IDebeUse<T> {
    const proxy = this;
    return new Proxy<any>(
      {},
      {
        get: function(t: string, methodName: string) {
          return (...args: [any]) => {
            return proxy[methodName](collection, ...args);
          };
        }
      }
    );
  }
  listen<T = IGetItem>(
    collection: string,
    callback: IObserverCallback<(T & IGetItem)[]>
  ): IUnlisten {
    return this.dispatcher.listen('insert', callback, { collection });
  }
  insert<T = IInsertItem>(
    collection: string,
    value: (T & IInsertItem)[],
    options?: IInsertInput
  ): Promise<(T & IGetItem)[]>;
  insert<T = IInsertItem>(
    collection: string,
    value: T & IInsertItem,
    options?: IInsertInput
  ): Promise<T & IGetItem>;
  insert<T = IInsertItem>(
    collection: string,
    value: (T & IInsertItem)[] | (T & IInsertItem),
    optionsInput: IInsertInput = {}
  ): Promise<(T & IGetItem)[] | T & IGetItem> {
    const options: IInsert = {
      refetchResult: false,
      update: true,
      ...optionsInput
    } as any;
    if (Array.isArray(value)) {
      return this.dispatcher.run<T & IGetItem>(
        'insert',
        collection,
        value,
        options
      );
    }
    return this.dispatcher
      .run<(T & IGetItem)[]>('insert', collection, [value], options)
      .then(([x]) => x);
  }
  // remove
  remove<T = any>(
    collection: string,
    value: string | string[]
  ): Promise<string | string[]> {
    if (Array.isArray(value)) {
      return this.dispatcher.run('remove', collection, value);
    }
    return this.dispatcher
      .run('remove', collection, [value])
      .then(x => x && x[0]);
  }
  // all
  all<T = TBase>(collection: string): Promise<(T & IGetItem)[]>;
  all<T = TBase>(
    collection: string,
    value: string | string[] | IQueryInput
  ): Promise<(T & IGetItem)[]>;
  all<T = TBase>(
    collection: string,
    callback: IObserverCallback<(T & IGetItem)[]>
  ): IUnlisten;
  all<T = TBase>(
    collection: string,
    value: string[] | string | IQueryInput,
    callback: IObserverCallback<(T & IGetItem)[]>
  ): IUnlisten;
  all<T = TBase>(
    collection: string,
    value?:
      | string
      | string[]
      | IQueryInput
      | IObserverCallback<(T & IGetItem)[]>,
    callback?: IObserverCallback<(T & IGetItem)[]>
  ): Promise<T[]> | IUnlisten {
    if (value && typeof value === 'function') {
      callback = value;
      value = undefined;
    }
    if (value && typeof value === 'string') {
      value = { id: value };
    } else if (value && Array.isArray(value)) {
      value = { id: value };
    } else if (value && typeof value === 'object' && Array.isArray(value.id)) {
      value = { id: value.id };
    }
    const query = ensureQuery(value);
    if (callback) {
      return this.dispatcher.listen<T[]>('all', callback, {
        collection,
        query
      });
    }
    return this.dispatcher.run<T[]>('all', collection, query);
  }
  // count
  count(collection: string): Promise<number>;
  count(
    collection: string,
    value: string | string[] | IQueryInput
  ): Promise<number>;
  count(collection: string, callback: IObserverCallback<number>): IUnlisten;
  count(
    collection: string,
    value: IQueryInput,
    callback: IObserverCallback<number>
  ): IUnlisten;
  count(
    collection: string,
    value?: string | string[] | IQueryInput | IObserverCallback<number>,
    callback?: IObserverCallback<number>
  ): Promise<number> | IUnlisten {
    if (value && typeof value === 'function') {
      callback = value;
      value = undefined;
    }
    if (value && typeof value === 'string') {
      value = { id: value };
    } else if (value && Array.isArray(value)) {
      value = { id: value };
    } else if (value && typeof value === 'object' && Array.isArray(value.id)) {
      value = { id: value.id };
    }
    const query = ensureQuery(value);
    if (callback) {
      return this.dispatcher.listen<number>('count', callback, {
        collection,
        query
      });
    }
    return this.dispatcher.run<number>('count', collection, query);
  }
  // get
  get<T = TBase>(
    collection: string,
    args?: IQueryInput | string
  ): Promise<T & IGetItem>;
  get<T = TBase>(
    collection: string,
    value?: IQueryInput | string,
    callback?: IObserverCallback<T & IGetItem>
  ): IUnlisten;
  get<T = TBase>(
    collection: string,
    value?: IQueryInput | string,
    callback?: IObserverCallback<T & IGetItem>
  ): Promise<T & IGetItem> | IUnlisten {
    if (value && typeof value === 'function') {
      callback = value;
      value = undefined;
    }
    if (
      value &&
      typeof value === 'object' &&
      value.id &&
      !Array.isArray(value.id)
    ) {
      value = value.id;
    }
    if (value && typeof value === 'string') {
      if (callback) {
        return this.dispatcher.listen<T>('get', callback, {
          collection,
          query: value
        });
      }
      return this.dispatcher.run<T & IGetItem>('get', collection, value);
    }
    if (value && Array.isArray(value)) {
      value = { id: value };
    } else if (value && typeof value === 'object' && Array.isArray(value.id)) {
      value = { id: value.id };
    }
    const query = ensureQuery(value);
    if (callback !== undefined) {
      const cb = callback;
      return this.dispatcher.listen<T[]>(
        'all',
        (error, value) => cb(error, error ? undefined : (value[0] as any)),
        {
          collection,
          query
        }
      );
    }
    return this.dispatcher
      .run<(T & IGetItem)[]>('all', collection, query)
      .then(x => x && x[0]);
  }
}
