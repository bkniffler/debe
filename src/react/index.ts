export * from './react';
export * from './provider';
export * from './sync';
export * from './context';
