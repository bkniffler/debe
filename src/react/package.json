{
  "name": "debe-react",
  "main": "index.ts",
  "version": "1.0.0",
  "dependencies": {
    "@types/redux": "^3.6.0",
    "debe": "*",
    "debe-socket": "*",
    "debe-sync": "*",
    "react-redux": "^7.1.0",
    "redux": "^4.0.1"
  },
  "devDependencies": {
    "@types/react": "^16.8.17",
    "@types/react-dom": "^16.8.4",
    "jest-dom": "^3.2.2",
    "react": "^16.9.0-alpha.0",
    "react-dom": "^16.9.0-alpha.0",
    "react-testing-library": "^7.0.0"
  }
}
