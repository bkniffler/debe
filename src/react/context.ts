import * as React from 'react';
import { Debe } from 'debe';

export const debeContext = React.createContext<Debe>(undefined as any);
