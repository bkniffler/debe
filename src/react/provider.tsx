import * as React from 'react';
import { Debe } from 'debe';
import {
  applyMiddleware,
  compose,
  createStore,
  combineReducers,
  Middleware,
  Reducer
} from 'redux';
// @ts-ignore
import { Provider } from 'react-redux';
import { useMemoOne } from './utils/memo-one';

export let STOREKEY = 'debe';
export let TIMEOUT = 5000;
export let setRequestTimeout = (timeout: number) => (TIMEOUT = timeout);
export let setStoreKey = (key: string) => (STOREKEY = key);

export const context = React.createContext<Debe>(undefined as any);
export function DebeProvider({
  children,
  initialize,
  loading,
  store: s,
  error,
  render,
  value
}: {
  store?: any;
  render?: (db: Debe) => React.ReactNode;
  loading?: () => React.ReactNode;
  error?: (err: any) => React.ReactNode;
  initialize?: (db: Debe) => Promise<void>;
  value: Debe | (() => Debe);
  children?: React.ReactNode;
}) {
  const [err, update] = React.useState<any>(undefined);
  const db = useMemoOne<Debe>(() => {
    const _db = value && (value as Debe).all ? value : (value as Function)();
    return _db;
  });
  const store = useMemoOne<any>(() => s || makeStore(db, {}));
  React.useEffect(() => {
    async function g() {
      try {
        if (!db.isInitialized) {
          await db.initialize();
        }
        if (initialize) {
          await initialize(db);
        }
        update(null);
      } catch (err) {
        update(err);
      }
    }
    g();
  }, [db]);

  if (render && db.isInitialized) {
    children = render(db);
  } else if (loading && !db.isInitialized) {
    children = loading();
  } else if (err && error) {
    children = error(err);
  }

  return (
    <context.Provider value={db}>
      <Provider store={store}>{children}</Provider>
    </context.Provider>
  );
}

export const reducer: Reducer = (state, action) => {
  if (!state) {
    return {};
  }
  switch (action.type) {
    case 'DEBE_QUERY_LOADING':
      return { ...state, [action.key]: { loading: true } };
    case 'DEBE_QUERY_SUCCESS':
      return {
        ...state,
        [action.key]: { loading: false, result: action.result }
      };
    case 'DEBE_QUERY_ERROR':
      return {
        ...state,
        [action.key]: { loading: false, error: action.error }
      };
  }
  return state;
};

const middleware = (db: Debe): Middleware => store => next => action => {
  if (action.type === 'DEBE_QUERY') {
    const { key, collection, query, method } = action;
    const stop = db[method as 'all'](collection, query, (error, result) => {
      if (timeout) {
        clearTimeout(timeout);
      }
      if (error) {
        next({ type: 'DEBE_QUERY_ERROR', error, key });
      } else {
        next({ type: 'DEBE_QUERY_SUCCESS', result, key });
      }
    });
    let timeout = setTimeout(() => {
      stop();
      next({
        type: 'DEBE_QUERY_ERROR',
        error: new Error('Query ' + key + ' timed out'),
        key
      });
    }, TIMEOUT);
    return next({ type: 'DEBE_QUERY_LOADING', key });
  } else if (action.type === 'DEBE_QUERY') {
    const { key, collection, query, method } = action;
    db[method as 'all'](collection, query)
      .then(result => {
        next({ type: 'DEBE_QUERY_SUCCESS', result, key });
      })
      .catch(error => {
        next({ type: 'DEBE_QUERY_ERROR', error, key });
      });
    return next({ type: 'DEBE_QUERY_LOADING', key });
  }
  return next(action);
};

export const makeStore = (db: Debe, initialState: any = {}) => {
  const composeEnhancers =
    (typeof window !== 'undefined' &&
      window['__REDUX_DEVTOOLS_EXTENSION_COMPOSE__']) ||
    compose;
  const rootReducer = combineReducers({
    debe: reducer
  });
  return createStore(
    rootReducer,
    initialState,
    composeEnhancers(applyMiddleware(middleware(db)))
  );
};

export function getUnloadedKeys(store: any, storeKey: string = STOREKEY) {
  const state = store.getState()[storeKey];
  const promises = [];
  for (let key of Object.keys(state)) {
    const entry = state[key];
    if (state[key].loading) {
      promises.push(entry);
    }
  }
  console.log(promises);
  return promises;
}

export async function awaitLoaded(
  store: any,
  unloaded?: any[],
  storeKey: string = STOREKEY,
  timeout: number = 20
): Promise<any> {
  if (!unloaded) {
    unloaded = getUnloadedKeys(store, storeKey);
  }
  if (unloaded.length === 0) {
    return;
  }
  await new Promise(yay => setTimeout(yay, timeout));
  return awaitLoaded(
    store,
    getUnloadedKeys(store, storeKey),
    storeKey,
    timeout
  );
}
