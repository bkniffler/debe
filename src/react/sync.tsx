import * as React from 'react';
import { debeContext } from './context';
import { Sync } from 'debe-sync';
import { IConnectionState } from 'asyngular-client';

export const syncContext = React.createContext<Sync>(undefined as any);

export function useConnectionState(): IConnectionState {
  const client = React.useContext(debeContext);
  const [state, setState] = React.useState<IConnectionState>(
    client['sync'] && client['sync'].state
  );
  React.useEffect(() => {
    if (!client || !client['sync'] || !client['sync'].onConnectionState) {
      return;
    }
    return client['sync'].onConnectionState((con: IConnectionState) =>
      setState(con)
    );
  }, [client]);
  return state;
}
