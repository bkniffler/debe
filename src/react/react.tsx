import * as React from 'react';
import { IDebeUse, IItem, IQueryInput } from 'debe';
import { useMemoOne } from './utils/memo-one';
// @ts-ignore
import { useDispatch, useSelector } from 'react-redux';
import { context } from './provider';

const empty = {};
function useDebeBase(
  method: 'get' | 'all',
  collection: string,
  query: (IQueryInput & { skip?: boolean }) | string | undefined,
  defaultValue: any,
  once = false
) {
  const skip = query && query['skip'];
  if (query) {
    delete query['skip'];
  }
  const dispatch = useDispatch();
  const str = skip ? '' : JSON.stringify(query || {});
  const key = useMemoOne<string>(() => {
    return skip ? '' : `${collection}:${method}:${str}`;
  }, [collection, str, skip]);
  const { result = defaultValue, error, loading } = useSelector((store: any) =>
    !key || !store.debe[key] ? empty : store.debe[key]
  );
  const id = useMemoOne(() => fastGuid());
  useMemoOne<string>(() => {
    if (!key) {
      return;
    }
    dispatch({
      id,
      key,
      type: 'DEBE_QUERY',
      collection,
      query,
      method,
      once
    });
  }, [key]);
  React.useEffect(() => {
    if (!once && !key) {
      return () => {
        dispatch({
          id,
          key,
          type: 'DEBE_QUERY_UNLISTEN'
        });
      };
    }
    return undefined;
  }, [key]);
  return [result, loading, error];
}

export function useAllOnce<T>(
  collection: string,
  query?: (IQueryInput & { skip?: boolean }) | string[]
) {
  return useAll(collection, query, true);
}

export function useGetOnce<T>(
  collection: string,
  query?: (IQueryInput & { skip?: boolean }) | string
) {
  return useGet(collection, query, true);
}

const emptyArray: any[] = [];
export function useAll<T>(
  collection: string,
  query?: (IQueryInput & { skip?: boolean }) | string[],
  once = false
) {
  return useDebeBase('all', collection, query as any, emptyArray, once);
}

export function useGet<T>(
  collection: string,
  query?: (IQueryInput & { skip?: boolean }) | string,
  once = false
) {
  return useDebeBase('get', collection, query, undefined, once);
}

export function useCollection<T = IItem>(service: string): IDebeUse<T> {
  const db = React.useContext(context);
  return useMemoOne(() => (db && db.use(service)) as any, [db]);
}

export function createUse<T = IItem>(collection: string) {
  return {
    useAll: function(query?: (IQueryInput & { skip?: boolean }) | string[]) {
      return useAll<T>(collection, query);
    },
    useAllOnce: function(
      query?: (IQueryInput & { skip?: boolean }) | string[]
    ) {
      return useAllOnce<T>(collection, query);
    },
    useGet: function(query?: (IQueryInput & { skip?: boolean }) | string) {
      return useGet<T>(collection, query);
    },
    useGetOnce: function(query?: (IQueryInput & { skip?: boolean }) | string) {
      return useGetOnce<T>(collection, query);
    },
    useCollection: function() {
      return useCollection<T>(collection);
    }
  };
}

var lut: string[] = [];
for (var i = 0; i < 256; i++) {
  lut[i] = (i < 16 ? '0' : '') + i.toString(16);
}
function fastGuid() {
  var d0 = (Math.random() * 0xffffffff) | 0;
  var d1 = (Math.random() * 0xffffffff) | 0;
  var d2 = (Math.random() * 0xffffffff) | 0;
  var d3 = (Math.random() * 0xffffffff) | 0;
  return (
    lut[d0 & 0xff] +
    lut[(d0 >> 8) & 0xff] +
    lut[(d0 >> 16) & 0xff] +
    lut[(d0 >> 24) & 0xff] +
    '-' +
    lut[d1 & 0xff] +
    lut[(d1 >> 8) & 0xff] +
    '-' +
    lut[((d1 >> 16) & 0x0f) | 0x40] +
    lut[(d1 >> 24) & 0xff] +
    '-' +
    lut[(d2 & 0x3f) | 0x80] +
    lut[(d2 >> 8) & 0xff] +
    '-' +
    lut[(d2 >> 16) & 0xff] +
    lut[(d2 >> 24) & 0xff] +
    lut[d3 & 0xff] +
    lut[(d3 >> 8) & 0xff] +
    lut[(d3 >> 16) & 0xff] +
    lut[(d3 >> 24) & 0xff]
  );
}
