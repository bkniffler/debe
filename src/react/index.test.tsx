import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { useAll, DebeProvider, useCollection } from './index';
import { render, waitForElement, fireEvent } from 'react-testing-library';
import { MemoryAdapter } from 'debe-memory';
import 'jest-dom/extend-expect';
import { Debe } from 'debe';

interface ILorem {
  name: string;
}
const collections = [{ name: 'lorem', index: ['name'] }];
test('react:basic', async cb => {
  function Component() {
    const [result] = useAll<ILorem>('lorem', {});
    return (
      <span data-testid="result">{result.map(x => x.name).join(', ')}</span>
    );
  }

  const { getByTestId, asFragment } = render(
    <DebeProvider
      initialize={async db => {
        await db.insert('lorem', [{ name: 'Beni' }, { name: 'Alex' }]);
      }}
      value={() => new Debe(new MemoryAdapter(), collections)}
      render={() => <Component />}
      loading={() => <span>Loading...</span>}
    />
  );

  await (act as any)(async () => {
    await waitForElement(() => getByTestId('result'));
  });

  //expect(resultNode).toHaveTextContent();
  expect(asFragment()).toMatchSnapshot();
  cb();
});

test('react:many', async cb => {
  function Component() {
    const [result] = useAll<ILorem>('lorem', {
      where: ['name < ?', '500']
    });
    return (
      <ul data-testid="result">
        <li>Start</li>
        {result.map(item => (
          <li key={item.id}>
            <span data-testid={item.name}>{item.name}</span>
          </li>
        ))}
        <li>End</li>
      </ul>
    );
  }

  const { getByTestId, asFragment } = render(
    <DebeProvider
      initialize={async db => {
        const items = [];
        for (let x = 0; x < 1000; x++) {
          items.push({ name: `${x}`.padStart(3, '0') });
        }
        await db.insert('lorem', items);
      }}
      value={() => new Debe(new MemoryAdapter(), collections)}
      render={() => <Component />}
      loading={() => <span>Loading...</span>}
    />
  );

  await (act as any)(async () => {
    await waitForElement(() => getByTestId('result'));
  });

  expect(() => getByTestId('500')).toThrow();
  expect(getByTestId('499')).toBeTruthy();
  expect(asFragment()).toMatchSnapshot();
  cb();
});

test('react:listen', async cb => {
  let renders = 0;
  function Component() {
    const [result] = useAll<ILorem>('lorem', {});
    renders = renders + 1;
    return (
      <ul>
        {result.map(item => (
          <li key={item.id}>
            <span data-testid={item.name}>{item.name}</span>
          </li>
        ))}
      </ul>
    );
  }

  const { getByTestId, asFragment } = render(
    <DebeProvider
      initialize={async db => {
        await db.insert('lorem', [{ name: 'Beni' }, { name: 'Alex' }]);
        setTimeout(
          () => db.insert('lorem', [{ name: 'Max' }, { name: 'Nik' }]),
          1000
        );
      }}
      value={() => new Debe(new MemoryAdapter(), collections)}
      render={() => <Component />}
      loading={() => <span>Loading...</span>}
    />
  );

  await (act as any)(async () => {
    await waitForElement(() => getByTestId('Nik'));
  });

  //expect(resultNode).toHaveTextContent();
  expect(renders).toBe(3);
  expect(asFragment()).toMatchSnapshot();
  cb();
});

test('react:interact', async cb => {
  function Component() {
    const collection = useCollection<ILorem>('lorem');
    const [result] = useAll<ILorem>('lorem', {});
    return (
      <div>
        <button
          data-testid="button"
          onClick={() => collection.insert({ name: 'Jadda' })}
        >
          Add
        </button>
        <ul>
          {result.map(item => (
            <li key={item.id}>
              <span data-testid={item.name}>{item.name}</span>
            </li>
          ))}
        </ul>
      </div>
    );
  }

  const { getByTestId, asFragment } = render(
    <DebeProvider
      initialize={async db => {
        await db.insert('lorem', [{ name: 'Beni' }, { name: 'Alex' }]);
      }}
      value={() => new Debe(new MemoryAdapter(), collections)}
      render={() => <Component />}
      loading={() => <span>Loading...</span>}
    />
  );

  await (act as any)(async () => {
    await waitForElement(() => getByTestId('Beni'));
  });

  //expect(resultNode).toHaveTextContent();
  await (act as any)(async () => {
    fireEvent.click(getByTestId('button'));
    await waitForElement(() => getByTestId('Jadda'));
  });

  expect(asFragment()).toMatchSnapshot();
  cb();
});
