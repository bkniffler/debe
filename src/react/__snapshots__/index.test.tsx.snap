// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react:basic 1`] = `
<DocumentFragment>
  <span>
    Beni, Alex
  </span>
</DocumentFragment>
`;

exports[`react:basic 2`] = `
Object {
  "lorem:all:undefined": Array [
    Object {
      "name": "Beni",
    },
    Object {
      "name": "Alex",
    },
  ],
}
`;

exports[`react:basic2 1`] = `
<DocumentFragment>
  <span>
    Beni, Alex
  </span>
</DocumentFragment>
`;

exports[`react:basic2 2`] = `
Object {
  "lorem:all:undefined": Array [
    Object {
      "name": "Beni",
    },
    Object {
      "name": "Alex",
    },
  ],
}
`;

exports[`react:suspense 1`] = `
<DocumentFragment>
  <span>
    Beni, Alex
  </span>
</DocumentFragment>
`;
